Run CI (Build Image and committer PRs) on self-hosted runner

ashb opened the pull request:
The queue for GitHub actions is starting to get painful.

This PR changes it to run any build of main (push or scheduled), or a PR
from a committer on the self-hosted runner.

However since this setting is configured _in_ the repo, it could be
changed by a PR, so we can't rely on that; we also use a custom build of
the self-hosted runner binary that enforces PRs are from contributors,
otherwise it will fail the build


<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

ashb committed:
Run CI (Build Image and committer PRs) on self-hosted runner

The queue for GitHub actions is starting to get painful.

This PR changes it to run any build of main (push or scheduled), or a PR
from a committer on the self-hosted runner.

However since this setting is configured _in_ the repo, it could be
changed by a PR, so we can't rely on that; we also use a custom build of
the self-hosted runner binary that enforces PRs are from contributors,
otherwise it will fail the build

potiuk commented on .github/workflows/ci.yml:
```diff
@@ -63,7 +62,13 @@ jobs:
   build-info:
     name: "Build info"
-    runs-on: ubuntu-20.04
+    runs-on: >-
```
I think this can be replaced by a pre-requisite step (before build-info) that could produce a string output that will be used as "fromJson()" in the "runs-on". I can make a fixup with it.

ashb replied:
I thought about that but there is one major downside: that would have to run on "public" queues, meaning it would still be at the mercy of long queue time

potiuk replied:
Not really. See https://github.com/astronomer/airflow/pull/1197

ashb replied:
Ohhh I see, yeah cool.

ashb replied:
Done in https://github.com/apache/airflow/pull/13730/commits/939a7fe948f2d917d1c1eca5100a1161cd5ab1b3

I kept it in the "Build Info" job, as each job needs a new worker, so having a worker tied up to run a very simple shell script seemed wasteful, especially as this falls reasonably cleanly under the "work out build info" description of the existing step

potiuk replied:
Yep. Easier indeed.

ashb committed:
Don't free space on self-hosted runners

It deletes the images we have pre-cached (defeating the point of
pre-loading them) and we manage freeing space via another process.

The exception to this is the K8s tests, where we don't need the images
we pre-cache, but we do want the space in our tmpfs.

ashb committed:
Some jobs take longer on self-hosted runners

ashb committed:
Work around supurious blocking IO failures

This seems to be an occasional problem on the build provider package job
which produces a lot of output all at once.

The error appears as the job just failing with an exit code but no
error.

potiuk commented on .github/workflows/ci.yml:
```diff
@@ -105,7 +117,12 @@ jobs:
       pullRequestLabels: ${{ steps.source-run-info.outputs.pullRequestLabels }}
+      runsOn: ${{ steps.set-runs-on.outputs.runsOn }}
     steps:
+      # Avoid having to specify the runs-on logic every time
```
Nice

potiuk reviewed (approved):
Looks good, except the fcntl part which likely can be removed for the in-memory run.

ashb committed:
When running KinD tests on CI, use pass through the host docker creds to Kind

This enabled self-hosted runners (which are logged in to a docker hub
user) to pull images without running in to pull limits.
